store/postgres/src/store.rs

```diff
@@ -1073,47 +1121,59 @@ impl StoreTrait for Store {
 
     fn block_number(
         &self,
-        _: &SubgraphDeploymentId,
+        subgraph_id: &SubgraphDeploymentId,
         hash: H256,
     ) -> Result<Option<BlockNumber>, StoreError> {
         use crate::db_schema::ethereum_blocks::dsl;
 
-        let hash = format!("{:x}", hash);
-
         // We should also really check that the block with the given hash is
         // on the chain starting at the subgraph's current head. That check is
         // very expensive though with the data structures we have currently
         // available. Ideally, we'd have the last REORG_THRESHOLD blocks in
         // memory so that we can check against them, and then mark in the
         // database the blocks on the main chain that we consider final
-        let number: Option<i64> = dsl::ethereum_blocks
-            .select(dsl::number)
-            .filter(dsl::hash.eq(hash))
-            .filter(dsl::network_name.eq(&self.network_name))
+        let block: Option<(i64, String)> = dsl::ethereum_blocks
+            .select((dsl::number, dsl::network_name))
+            .filter(dsl::hash.eq(format!("{:x}", hash)))
             .first(&*self.get_conn()?)
             .optional()?;
-        number
-            .map(|number| {
-                BlockNumber::try_from(number)
-                    .map_err(|e| StoreError::QueryExecutionError(e.to_string()))
+        let subgraph_network = self.network_name(subgraph_id)?;
+        block
+            .map(|(number, network_name)| {
+                if subgraph_network.is_none() || Some(&network_name) == subgraph_network.as_ref() {
+                    BlockNumber::try_from(number)
+                        .map_err(|e| StoreError::QueryExecutionError(e.to_string()))
+                } else {
+                    Err(StoreError::QueryExecutionError(format!(
+                        "subgraph {} belongs to network {} but block {:x} belongs to network {}",
+                        subgraph_id,
+                        subgraph_network.unwrap_or("(none)".to_owned()),
+                        hash,
+                        network_name
+                    )))
```

Comment:
Why bother checking the network? If the hash matches, then the number and all ancestor blocks are the same, and for any network that's not actively trying to mimick another one, the network will also be the same, not that it would matter that the networks are different since they are identical up to this point.

Reply:
It guards against mistakenly using a hash from one network against another network.

Reply:
To expand on that: if you use a hash that is not from the network in question, you should always get an error. But if that hash happens to be from another network that the node in question is also ingesting, we would resolve the hash to a number (according to the wrong network) and then accept the query. That introduces some reproducibility issues as whether such a query succeeds depends on whether an indexer is also indexing from the other network or not.

Reply:
Oh I see now, thanks for explaining. The network will be included in the block parameter in the v2 API which means we won't need to infer it from the subgraph, but for now this is a good solution.